### ISSUE TITLE
groups: make group all an alias of @*

### ISSUE THREAD

**Issue by thiell** [cea-hpc/clustershell#335]
Group `all` and `@*` should be the same to avoid confusion.

**Comment by jbaksta** [comment]
Just curious on this idea here. But also, maybe some consideration that I've not yet tested whether or not you can do or dug in deep enough.

Can `@group:*` be overridden in a configuration file to set what would be the default for `@group:*`?

Furthermore, I'd be curious if behavior of `@group` without the `:*` could behave as if `:*` was implied when there are subgroups. I like the wildcard artifact here as explicit, but I'd also just personally ack that if I leave off the subgroup name, I'd presume I want everything (or if I override what is `:*` in configuration in prior question). I'd also like to not have what might be unnecessary '*' to escape in various scripts (presently with a configuration file of `:all`), but it's no show stopper by any means; simply what I'd consider a quality of life feature.

**Comment by jbaksta** [comment]
My reading of the docs a couple more times and some tinkering do demonstrate that `@group:all` does match that the `@group:*` expands to what is defined by `@group:all`, so that's good and I didn't understand that before when I implemented all. So I think that only leaves the second question about implicit `:all` or `:*`.

**Comment by thiell** [comment]
Thanks @jbaksta for digging in! This is great feedback. To confirm your first question: yes, what `@source:*` (and `clush -a` / `nodeset -a`) expands to is configurable per source: for YAML/file sources it's the special `all` group if defined (otherwise the union of all groups), and for upcall sources it's the `all` upcall (otherwise the union via `list` + `map`, or the new `mapall` in 1.10+). Hopefully the doc is clear enough at https://clustershell.readthedocs.io/en/latest/config.html#node-groups but we can always improve it (if you have ideas that would be welcome).

I played with AI with an isolated configs via `CLUSTERSHELL_CFGDIR`; `n0` = adm node, `n[1-4]` = compute, all = compute when defined, and I believe today we have this:
```
  ┌────────────────────────────────────────────┬────────┬────────┬────────┐
  │                   Config                   │   @*   │  @all  │   -a   │
  ├────────────────────────────────────────────┼────────┼────────┼────────┤
  │ flat-file with all: line (shipped default) │ n[1-4] │ n[1-4] │ n[1-4] │
  ├────────────────────────────────────────────┼────────┼────────┼────────┤
  │ YAML with all group                        │ n[1-4] │ n[1-4] │ n[1-4] │
  ├────────────────────────────────────────────┼────────┼────────┼────────┤
  │ YAML without all group                     │ n[0-4] │ empty  │ n[0-4] │
  ├────────────────────────────────────────────┼────────┼────────┼────────┤
  │ upcall source, no all upcall (slurm-style) │ n[0-4] │ empty  │ n[0-4] │
  └────────────────────────────────────────────┴────────┴────────┴────────┘
```

Would you agree that we can summarize the need as: make `@all` / `@source:all` resolve exactly like `@*` / `@source:*` (and `nodeset`/`clush -a`) whenever the source doesn't define a group named `all` ( instead of silently expanding to an empty node set). Since that turns today's empty result into "all nodes", it's a behavior change, but always from an empty result to something (which I think would be acceptable for 1.11). While testing I also noticed that a group named `all` defined in `mapall` output is currently not honored by `-a` / `@source:*` (the union of all groups wins, `all` is included as a regular member), unlike YAML sources. Probably something to fix alongside!

**Comment by jbaksta** [comment]
Yes, I think that would be preferred. I think to clarify a bit, what I was curious about is the following end result:

```
> cluset -e @x9000

> cluset -e @x9000:all
guc0001 guc0002 guc0003 guc0004 guc0005 guc0006 guc0007 guc0008 guc0009 guc0010 guc0011 guc0012 guc0013 guc0014 guc0015 guc0016 gug0001 gug0002
> cluset -e @x9000:\*
guc0001 guc0002 guc0003 guc0004 guc0005 guc0006 guc0007 guc0008 guc0009 guc0010 guc0011 guc0012 guc0013 guc0014 guc0015 guc0016 gug0001 gug0002

> cluset -f @x9000

> cluset -f @x9000:all
guc[0001-0016],gug[0001-0002]
> cluset -f @x9000:\*
guc[0001-0016],gug[0001-0002]
```

where I think in the absence of explicitly typing `:all` or `:*`, when it does exists, it should expand to the special `all`. So:

```
> cluset -f @x9000
guc[0001-0016],gug[0001-0002]
> cluset -f @x9000:all
guc[0001-0016],gug[0001-0002]
> cluset -f @x9000:\*
guc[0001-0016],gug[0001-0002]
```

**Comment by thiell** [comment]
This makes sense to me now, thank you for the clarification.
Currently, the group syntax is `@source:group` and `@group` is just a convenience form for resolving groups the **default** group source. So the problem is how would you differentiate the **group** named `@x9000` in the default group source from the group **source** named `@x9000:` that is normally used as `@x9000:<groupname>`? We _could_ accept the risk of conflict with a priority but I'm a bit worried that could confuse some users.

@degremont fyi, let me know what you think when you're back from vacation

**Comment by thiell** [comment]
@jbaksta I also wanted to add that the following is already available today:
```
cluset -s x9000 -f -a         # fold all from source x9000 →  guc[0001-0016],gug[0001-0002]
clush  -s x9000 -a  <cmd>
```

Related:  `@x9000:` with an empty group name is already meaningful, it passes an empty `$GROUP` to the upcall and some setups rely on that, so it can't be repurposed to mean `:all`. I opened #705 so a source can supply its own fallback in that case (e.g. `${GROUP:-.*}` when a "all regex" is wanted).

Finally, I'll see what can be done to align `@all` with `@*` by default, but I don't think we'll go further than that for now, a bare `@x9000` would be ambiguous with a group actually named `x9000` in the default source. Hope that makes sense!
